tests/tls_handler_test.c

```diff
@@ -0,0 +1,699 @@
+    ASSERT_INT_EQUALS(1, incoming_rw_args.read_invocations);
+
+    /* Go ahead and verify back-pressure works*/
+    rw_handler_trigger_increment_read_window(incoming_args.rw_handler, incoming_args.rw_slot, 100);
+    rw_handler_trigger_increment_read_window(outgoing_args.rw_handler, outgoing_args.rw_slot, 100);
+
+    ASSERT_SUCCESS(
+        aws_condition_variable_wait_pred(&condition_variable, &mutex, s_tls_test_read_predicate, &incoming_rw_args));
+    ASSERT_SUCCESS(
+        aws_condition_variable_wait_pred(&condition_variable, &mutex, s_tls_test_read_predicate, &outgoing_rw_args));
+
+    ASSERT_INT_EQUALS(2, outgoing_rw_args.read_invocations);
+    ASSERT_INT_EQUALS(2, incoming_rw_args.read_invocations);
+
+    ASSERT_BIN_ARRAYS_EQUALS(
+        write_tag.buffer,
+        write_tag.len,
+        incoming_rw_args.received_message.buffer,
+        incoming_rw_args.received_message.len);
+    ASSERT_BIN_ARRAYS_EQUALS(
+        read_tag.buffer, read_tag.len, outgoing_rw_args.received_message.buffer, outgoing_rw_args.received_message.len);
+
+    aws_channel_shutdown(incoming_args.channel, AWS_OP_SUCCESS);
+    ASSERT_SUCCESS(aws_condition_variable_wait_pred(
+        &condition_variable, &mutex, s_tls_channel_shutdown_predicate, &incoming_args));
+
+    /*no shutdown on the client necessary here (it should have been triggered by shutting down the other side). just
+     * wait for the event to fire. */
+    ASSERT_SUCCESS(aws_condition_variable_wait_pred(
+        &condition_variable, &mutex, s_tls_channel_shutdown_predicate, &outgoing_args));
+
+    aws_client_bootstrap_clean_up(&client_bootstrap);
+    ASSERT_SUCCESS(aws_server_bootstrap_destroy_socket_listener(&server_bootstrap, listener));
+    aws_server_bootstrap_clean_up(&server_bootstrap);
+    aws_tls_ctx_destroy(client_ctx);
+    aws_tls_ctx_destroy(server_ctx);
+
+    aws_event_loop_group_clean_up(&el_group);
+    aws_tls_clean_up_static_state();
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(tls_channel_echo_and_backpressure_test, s_tls_channel_echo_and_backpressure_test_fn)
+
+struct default_host_callback_data {
+    struct aws_host_address aaaa_address;
+    struct aws_host_address a_address;
+    bool has_aaaa_address;
+    bool has_a_address;
+    struct aws_condition_variable condition_variable;
+    bool invoked;
+};
+
+static int s_verify_negotiation_fails(struct aws_allocator *allocator, const struct aws_string *host_name) {
+
+    aws_tls_init_static_state(allocator);
+
+    struct aws_event_loop_group el_group;
+    ASSERT_SUCCESS(aws_event_loop_group_default_init(&el_group, allocator, 0));
+
+    struct aws_mutex mutex = AWS_MUTEX_INIT;
+    struct aws_condition_variable condition_variable = AWS_CONDITION_VARIABLE_INIT;
+
+    struct aws_tls_ctx_options client_ctx_options;
+    aws_tls_ctx_options_init_default_client(&client_ctx_options);
+    aws_tls_ctx_options_override_default_trust_store(&client_ctx_options, "/etc/ssl/certs", NULL);
```

Comment:
What's this going to do on windows?

Reply:
nothing. It will probably break on non debian linux distributions though. We need to add some path detection stuff to pki utils. Let's move that to a separate review though.

Reply:
👍